core/worker.go

```diff
@@ -132,54 +128,38 @@ func (p *WorkerPool) runWorker(st *stats.Stats, idlingCounter *int32, id int) {
 				return
 			}
 			setWorking()
-
-			tries := 0
-			for job != nil {
-				err := job.Run(&wp)
-				var acceptableErr AcceptableError
-				if err != nil {
-					acceptableErr = IsAcceptableError(err)
-					if acceptableErr != nil {
-						err = nil
-					}
-				}
-
-				if err != nil {
-					errCode, doRetry := IsRetryableError(err)
-					if doRetry && p.params.Retries > 0 && tries < p.params.Retries {
-						tries++
-						sleepTime := bkf.NextBackOff()
-						log.Printf(`?%s "%s", sleep for %v`, errCode, job, sleepTime)
-						select {
-						case <-time.After(sleepTime):
-							wp.st.Increment(stats.RetryOp)
-							continue
-						case <-p.ctx.Done():
-							run = false // if Canceled during sleep, report ERR and immediately process failCommand
-						}
-					}
-
-					job.PrintErr(err)
-					wp.st.Increment(stats.Fail)
-					job.Notify(false)
-					job = job.failCommand
-				} else {
-					if acceptableErr != ErrDisplayedHelp {
-						job.PrintOK(acceptableErr)
-					}
-					job.Notify(true)
-					job = job.successCommand
+			for {
+				job = p.runJob(wp, job)
+				if job == nil {
+					return
 				}
-				tries = 0
-				bkf.Reset()
 			}
-
 		case <-p.ctx.Done():
 			return
 		}
 	}
 }
 
+// runJob runs parent job and returns sub-job.
+func (p *WorkerPool) runJob(wp WorkerParams, job *Job) *Job {
+	if err := job.Run(&wp); err != nil {
```

Comment:
let's handle the non-error case first to reduce the indentation.

Reply:
Logic was moved into `job.go` and checks non-error case first.